### PR TITLE
Added support for startDate and endDate in repeatEvery

### PR DESCRIPTION
**Pull request by designerkamal**
Based on #361 
```
job.repeatEvery('0 6 * * *', {
  startDate: new Date('2018-05-01T00:00:00.000Z'),
  endDate: new Date('2018-05-31T00:00:00.000Z')
});
// runs every day at 6pm from May 1 - May 15
job.save();
```
- Added tests
- Updated Readme

Also, added following rule to `xo`
```
"no-unused-vars": [
        "error",
        {
          "argsIgnorePattern": "^_"
        }
]
```
to get rid of that one linter error in `createIndex` callback. I am sure this will also be useful in the future in similar cases.

### CODE REVIEW

**Comment by simison** [comment]
Whoops, this needs rebasing after promise rewrite.

Case example: https://github.com/agenda/agenda/issues/544

File: test/job.js [hunk 1]
```diff
@@ -94,6 +94,20 @@ describe('Job', () => {
...
     });
+    it('sets the startDate', () => {
+      const _start = new Date('2018-01-01T06:00:00-00:00');
+      job.repeatEvery(5000, {
+        startDate: _start
+      });
+      expect(job.attrs.startDate.valueOf()).to.be(_start.valueOf());
+    });
+    it('sets the endDate', () => {
+      const _end = new Date('2018-01-01T06:00:00-00:00');
```
Comment (wingsbob) [inline test/job.js]: Why the `_` prefix?

File: test/job.js [hunk 2]
```diff
@@ -212,6 +226,78 @@ describe('Job', () => {
...
+      const _start = new Date('2015-01-03');
+      job.attrs.lastRunAt = now;
+      job.repeatEvery('0 6 * * *', {
+        timezone: 'GMT',
+        startDate: _start
+      });
+      job.computeNextRunAt();
+      expect(job.attrs.nextRunAt.valueOf()).to.greaterThan(_start.valueOf());
+    });
+
```
Comment (wingsbob) [inline test/job.js]: Don't think we've got the empty lines between tests elsewhere

Comment (wingsbob) [reply test/job.js]: Ok, it does seem to chop and change 😐

File: lib/job/compute-next-run-at.js [hunk 3]
```diff
@@ -38,21 +42,67 @@ module.exports = function() {
...
-      } catch (e) {}
+
+        debug('[%s:%s] nextRunAt set to [%s]', this.attrs.name, this.attrs._id, this.attrs.nextRunAt ? this.attrs.nextRunAt.toISOString() : null);
+      } catch (e) { }
     } finally {
-      if (isNaN(this.attrs.nextRunAt)) {
+      if (this.attrs.nextRunAt !== undefined && isNaN(this.attrs.nextRunAt)) {
         this.attrs.nextRunAt = undefined;
         debug('[%s:%s] failed to calculate nextRunAt due to invalid repeat interval', this.attrs.name, this.attrs._id);
         this.fail('failed to calculate nextRunAt due to invalid repeat interval');
```
Comment (wingsbob) [inline lib/job/compute-next-run-at.js]: Do we have a test that we don't fail when the job has reached the end date?

File: lib/job/compute-next-run-at.js [hunk 4]
```diff
@@ -38,21 +42,67 @@ module.exports = function() {
...
+            dateOffset = cronTime._getNextDateFrom(dateOffset.valueOf());
+          }
+          nextDate = dateOffset;
+        }
+      }
+      // If startDate exists, nextDate is moved after startDate
+      if (endDate !== null && nextDate > endDate) {
+        this.attrs.nextRunAt = undefined;
+        this.attrs.repeatInterval = undefined;
+        this.attrs.type = 'normal';
```
Comment (wingsbob) [inline lib/job/compute-next-run-at.js]: What does this do? 😕

File: test/job.js [hunk 5]
```diff
@@ -212,6 +226,78 @@ describe('Job', () => {
...
+    it('set at startDate when startDate is in future and interval is humanInterval ', () => {
+      const now = new Date('2015-01-01T06:00:00-00:00');
+      const _start = new Date('2015-01-03');
+      job.attrs.lastRunAt = now;
+      job.repeatEvery('5 minutes', {
+        timezone: 'GMT',
+        startDate: _start
+      });
+      job.computeNextRunAt();
+      expect(job.attrs.nextRunAt.valueOf()).to.be(_start.valueOf());
```
Comment (wingsbob) [inline test/job.js]: I have a suspicion that this is hiding an issue where `job.attrs.nextRunAt` is now a number, and not a date.

Could all the date comparisons change to use `expect(dateObject).to.eql(new Date('expected-date-string'))`?

**Comment by leonardlin** [comment]
closing because it was implemented with https://github.com/agenda/agenda/pull/361

**Pull request closed without merging**